### ISSUE TITLE
1.3.46 regression: YANGSON doesn't correctly parse "when"  

### ISSUE THREAD

**Issue by andrew-lee-metaswitch** [CZ-NIC/yangson#34]
Using 1.3.45 I am able to validate a YAML file, but using 1.3.47 YANGSON (released today) it errors:

Validating sample/sdf-rpas.yaml..

> Traceback (most recent call last):
>   File "build/validate.py", line 17, in <module>
>     instance.validate()
>   File "/data/developer/.virtualenvs/AUT-949/lib/python3.6/site-packages/yangson/instance.py", line 336, in validate
>     self.schema_node._validate(self, scope, ctype)
>   File "/data/developer/.virtualenvs/AUT-949/lib/python3.6/site-packages/yangson/schemanode.py", line 485, in _validate
>     inst._member(m).validate(scope, ctype)
>   File "/data/developer/.virtualenvs/AUT-949/lib/python3.6/site-packages/yangson/instance.py", line 336, in validate
>     self.schema_node._validate(self, scope, ctype)
>   File "/data/developer/.virtualenvs/AUT-949/lib/python3.6/site-packages/yangson/schemanode.py", line 801, in _validate
>     super()._validate(inst, scope, ctype)
>   File "/data/developer/.virtualenvs/AUT-949/lib/python3.6/site-packages/yangson/schemanode.py", line 485, in _validate
>     inst._member(m).validate(scope, ctype)
>   File "/data/developer/.virtualenvs/AUT-949/lib/python3.6/site-packages/yangson/instance.py", line 336, in validate
>     self.schema_node._validate(self, scope, ctype)
>   File "/data/developer/.virtualenvs/AUT-949/lib/python3.6/site-packages/yangson/schemanode.py", line 977, in _validate
>     super()._validate(e, scope, ctype)
>   File "/data/developer/.virtualenvs/AUT-949/lib/python3.6/site-packages/yangson/schemanode.py", line 801, in _validate
>     super()._validate(inst, scope, ctype)
>   File "/data/developer/.virtualenvs/AUT-949/lib/python3.6/site-packages/yangson/schemanode.py", line 485, in _validate
>     inst._member(m).validate(scope, ctype)
>   File "/data/developer/.virtualenvs/AUT-949/lib/python3.6/site-packages/yangson/instance.py", line 336, in validate
>     self.schema_node._validate(self, scope, ctype)
>   File "/data/developer/.virtualenvs/AUT-949/lib/python3.6/site-packages/yangson/schemanode.py", line 977, in _validate
>     super()._validate(e, scope, ctype)
>   File "/data/developer/.virtualenvs/AUT-949/lib/python3.6/site-packages/yangson/schemanode.py", line 801, in _validate
>     super()._validate(inst, scope, ctype)
>   File "/data/developer/.virtualenvs/AUT-949/lib/python3.6/site-packages/yangson/schemanode.py", line 485, in _validate
>     inst._member(m).validate(scope, ctype)
>   File "/data/developer/.virtualenvs/AUT-949/lib/python3.6/site-packages/yangson/instance.py", line 336, in validate
>     self.schema_node._validate(self, scope, ctype)
>   File "/data/developer/.virtualenvs/AUT-949/lib/python3.6/site-packages/yangson/schemanode.py", line 801, in _validate
>     super()._validate(inst, scope, ctype)
>   File "/data/developer/.virtualenvs/AUT-949/lib/python3.6/site-packages/yangson/schemanode.py", line 483, in _validate
>     self._check_schema_pattern(inst, ctype)
>   File "/data/developer/.virtualenvs/AUT-949/lib/python3.6/site-packages/yangson/schemanode.py", line 506, in _check_schema_pattern
>     "member-not-allowed", m)
> yangson.exceptions.SchemaError: [/msw-deployment:deployment/sites/0/vnfcs/0/vnfc-vsphere-options] config member-not-allowed: datastore

where `vnfc-vsphere-options` has the following container:

    container vnfc-vsphere-options {
      when "ancestor::sites/site-parameters/vim-configuration/type = 'vsphere'";
      description
        "vSphere-specific options for this VNFC";
      uses vsphere:vsphere-options-per-vnfc;
      uses vsphere:vsphere-options-only-per-vnfci {
        when "derived-from(../../type, 'util:standalone-product')";
      }
    }

and `datastore` is define only in `vsphere:vsphere-options-only-per-vnfci`

**Comment by andrew-lee-metaswitch** [comment]
Have verified this is also broken on 1.3.46, so it looks like https://github.com/CZ-NIC/yangson/commit/077b0b1f0bf7adb955a09be5132a536d37d3cf18 is suspect

**Comment by andrew-lee-metaswitch** [comment]
I should note type is a `identityref`, and in this case, the type is "rpas" where:

```
 identity rpas {
    base util:standalone-product;
    base util:final-product;
    description
      "Identifies a VNFC as a RPAS";
  }
```

**Comment by llhotka** [comment]
Identities should be matched using XPath functions `derived-from` and `derived-from-or-self`, see [sec. 10.4](https://tools.ietf.org/html/rfc7950#section-10.4) in RFC 7950.

Using string equality is brittle. See also https://tools.ietf.org/html/rfc8407#section-4.6.2

**Comment by dimbleby** [comment]
The identity _is_ being matched via `derived-from`:

>     when "derived-from(../../type, 'util:standalone-product')";

(I see that the example contained two `when`s, both called `type`, so the confusion is understandable!)

**Comment by andrew-lee-metaswitch** [comment]
This was my bad:

> _ when "derived-from(../../type, 'util:standalone-product')";

should have been

>  when "derived-from(../.type, 'util:standalone-product')";_

So I should have taken the failed validation as a sign. :-(
Worth noting pyang has just released 2.1.1 which fixes the same issue there.

And thanks for the module!
